thorlcr/graph/thgraphslave.cpp

```diff
@@ -1258,6 +1258,23 @@ bool CSlaveGraph::serializeStats(MemoryBuffer &mb)
 
     CRuntimeStatisticCollection stats(graphStatistics);
     stats.setStatistic(StNumExecutions, numExecuted);
+
+    // Track active spills
+    IGraphTempHandler *tempHandler = queryJob().queryTempHandler(false);
+    if (tempHandler)
+    {
+        offset_t activeSpillSize = 0;
+        graph_id gid = queryGraphId();
+        Owned<IFileUsageIterator> iter = tempHandler->getIterator();
+        ForEach(*iter)
+        {
+            CFileUsageEntry &entry = iter->query();
+            if (entry.queryGraphId()==gid)
```

Comment:
Is this a graphid or a subgraph id.  I'm not sure why this is filtering....

Reply:
I seems that ::queryGraphId() should be return graphId but it returns subgraphId.

The filtering is to avoid double counting open spill files.  Each subgraph reports the spill size for files it has created.

Reply:
CGraphBase objects (and it's derivatives) always correspond to subgraphs.
A [outer] graph is submitted to thor as a job, those graphs correspond to Thor CJobBase objects.
The [outer]graph name is retried from queryJob()->queryGraphName()

Reply:
Afaics, this isn't going to work as planned.
When this method (CSlaveGraph::serliazeStats) is called, it's going to be either during, or at the end of it's life.
i.e. registered spills that were created by this subgraph will not be considered after this subgraph is finished.

For example, take 3 subgraphs, in series, where sg1 spills 's1', sg2 consumes 's1' and creates a new spill ('s2'), which sg3 consumes.
The sg1 will serialize it's spill aggregate, the master will call the job::addGraphSpill and add to total and peak.
The 2nd cosumes 's1' (deletes it), before producing 's2'.
sg2 serializes it's spill aggregate, the master addGraphSpill adds to total (ok as total of everything spilt ever), but then considers that total as the new peak. But it's not as this subgraph's spill is now the only spill in play.

i.e. there's no calculation of the current usage which goes down as well as up afaics.

The spill total related to the current subgraph could be useful as well (for diving down into what contributed to the peak), but the overall spill usage usage/peak needs to be the aggregate of all current spills registered in the handler, which will fluctuate up and down.

Reply:
You could continue to aggregate per subgraph and publish as a separate stat., though if not needed now, I'd leave for another PR, since you can already dive into a subgraph and see this detail manually by looking at the spill activities (there's usually only 1 or 2).

For the total currently in use (registered) *for the current graph (thor job)*, I think you need to check that you're in the root subgraph (owner == null), and total all registred in the job temp handler and serialize that.
That is the normal case (the job temp handler holds all the spills), however there is another case, where there's a contained group of child graphs in a resultsGraph, it will have it's own temp handler (this is so it clears up when goes out of scope).
These will need totalling and serializing too.

So I think you'll need something like (in CSlaveGraph::serializeStats):
```
   if (!owner)
      queryJob().queryTempHandler()->serializeUsageStats(stats);
   else
   {
        IGraphTempHandler *tempHandler = queryTempHandler(false);
        if (tempHandler) // graphResults temp handler
            tempHandler->serizlieUsageStats(stats);
    }
```

(NB: in theory I think there could be grandchild resultgraph temp handlers).

These current totals will need totalling up on the master and seeing if > peak.
This will be the peak for graph, not the job overall.

e.g. graph1 may be the 3 subgraph sequence example in my previous comment and had a peak due to 's2', but the next thor job (graph2), may exceed this.
So need to track workunit peak, read it, and increase it at the end of the job if exceeded too.